### PR TITLE
Add native loader entry points

### PR DESCRIPTION
**Pull request by jackcmay**
#### Problem

Cross-program instruction invocation needs to pass invocation context to the loader (context which most native programs don't need and should not be exposed to.

But also:
- The `bpf_loader` is tangled with `runtime` and does not need to be
- The `native_loader` is entwined with the `message_processor`

#### Summary of Changes

- Add support for a native loader specific entry point (in this PR the entry points are the same as a native program's, that will change shortly)
- Remove the `runtime`'s dependency on the `bpf_loader`
- Untangle the `native_loader` from the `message_processor`
- Encapsulate `native loader stuff` inside `NativeLoader`

These changes also change the `bpf_loader` dependency of others from implicit to explicit and highlight where further cleanup/extraction could occur, for example, `core` is dependent on `bpf_loader ` when it probably doesn't need to be.

Breaking this PR's changes out of the cross-invocation PR allows the `bpf_loader` to depend on the `runtime` where the `process_instruction` mechanics live rather than breaking those mechanics out to the SDK or somewhere else.

Fixes #

### CODE REVIEW

**Comment by jackcmay** [comment]
@mvines FYI, any objections?

File: sdk/src/native_loader.rs [hunk 1]
```diff
@@ -1,13 +1,46 @@
...
+#[derive(Debug, Clone, Copy, Deserialize, Eq, FromPrimitive, Hash, PartialEq, Serialize)]
+pub enum Kind {
+    Program = 1,
+    Loader = 2,
+}
+
+#[macro_export]
+macro_rules! native_program_info(
+    ($name:expr) => (
+        $crate::native_loader::Info{
```
Comment (mvines) [inline sdk/src/native_loader.rs:20]:
```suggestion
        $crate::native_loader::Info {
```

File: sdk/src/native_loader.rs [hunk 2]
```diff
@@ -1,13 +1,46 @@
...
+            kind: $crate::native_loader::Kind::Program,
+            name: $name.to_string(),
+        }
+    )
+);
+#[macro_export]
+macro_rules! native_loader_info(
+    ($name:expr) => (
+        $crate::native_loader::Info{
```
Comment (mvines) [inline sdk/src/native_loader.rs:29]:
```suggestion
        $crate::native_loader::Info {
```

File: sdk/src/native_loader.rs [hunk 3]
```diff
@@ -1,13 +1,46 @@
...
 
 crate::declare_id!("NativeLoader1111111111111111111111111111111");
 
+#[derive(Debug, Clone, Deserialize, Eq, Hash, PartialEq, Serialize)]
+pub struct Info {
+    pub kind: Kind,
+    pub name: String,
+}
+#[derive(Debug, Clone, Copy, Deserialize, Eq, FromPrimitive, Hash, PartialEq, Serialize)]
+pub enum Kind {
```
Comment (mvines) [inline sdk/src/native_loader.rs:12]: The name of this file, and then the `Kind` enum, trips me up.   This file appears to be about `native_loader`s and there are two `Kind`s of them: `Program` and `Loader`.


Perhaps rename the file to `native_program.rs` and then there are two `Kinds` of native programs: a normal `Kind::Program`, and `Kind::Loader` that can load other problems.   Does that seem to make sense?

File: sdk/src/native_loader.rs (hunk 2 above)
Comment (jackcmay) [reply sdk/src/native_loader.rs:29]: Thanks, it sucks that `fmt` doesn't always fixup macros

File: sdk/src/native_loader.rs (hunk 3 above)
Comment (jackcmay) [reply sdk/src/native_loader.rs:12]: Kinda, This file is the sdk (public) component of the `native_loader`, thus is contains the program id definition and the helper to create a loadable account.  Are you suggesting that we move `info` and `kind` (or maybe just `kind`) out to a new file called `native_program.rs`?

Comment (mvines) [reply sdk/src/native_loader.rs:12]: Maybe I'm just unclear on when I'd use `Kind::Program` vs `Kind::Loaader`

Comment (jackcmay) [reply sdk/src/native_loader.rs:12]: They are by the macros in `sdk/src/native_loader.rs` and in the `NativeLoader` implementation in `runtime/src/native_loader.rs`.  For the SDK consumer, they are wrapped by the macros.  This brings me to a larger question, this file is one that should probably not be in the "SDK" and instead should be in a "common" (yuck) spot in the tree.

**Pull request merged**